Plugins/Reveal/Reveal.cpp

```diff
@@ -0,0 +1,128 @@
+#include "Reveal.hpp"
+#include <string>
+#include "API/CNWSCreature.hpp"
+#include "API/CNWSFaction.hpp"
+#include "API/Functions.hpp"
+#include "Services/PerObjectStorage/PerObjectStorage.hpp"
+#include "ViewPtr.hpp"
+
+
+
+using namespace NWNXLib;
+using namespace NWNXLib::API;
+
+static ViewPtr<Reveal::Reveal> g_plugin;
+
+//key names for Per Object Storage
+const std::string revealKey = "REVEAL";
+const std::string detectionKey = "DETECTION VECTOR";
+
+//constants for detection method
+
+const int NWNX_REVEAL_SEEN = 1;
+const int NWNX_REVEAL_HEARD = 0;
+
+NWNX_PLUGIN_ENTRY Plugin::Info* PluginInfo()
+{
+    return new Plugin::Info
+    {
+        "Reveal",
+        "Functions to allow the selective revealing of a stealthed character to another character or their party.",
+        "Silvard",
+        "jusenkyo at gmail.com",
+        1,
+        true
+    };
+}
+
+NWNX_PLUGIN_ENTRY Plugin* PluginLoad(Plugin::CreateParams params)
+{
+    g_plugin = new Reveal::Reveal(params);
+    return g_plugin;
+}
+
+namespace Reveal {
+
+Reveal::Reveal(const Plugin::CreateParams& params)
+  : Plugin(params)
+{
+#define REGISTER(func)              \
+    GetServices()->m_events->RegisterEvent(#func, std::bind(&Reveal::func, this, std::placeholders::_1))
+
+    REGISTER(RevealTo);
+    REGISTER(SetRevealToParty);
+
+#undef REGISTER
+
+    GetServices()->m_hooks->RequestExclusiveHook<Functions::CNWSCreature__DoStealthDetection, int32_t,CNWSCreature*,CNWSCreature*, int32_t, int32_t*, int32_t*, int32_t>(&HookStealthDetection);
+    m_DoStealthDetection = GetServices()->m_hooks->FindHookByAddress(Functions::CNWSCreature__DoStealthDetection);
+}   
+
+Reveal::~Reveal()
+{
+}
+int32_t Reveal::HookStealthDetection(NWNXLib::API::CNWSCreature* pObserverCreature, NWNXLib::API::CNWSCreature* pHidingCreature, int32_t bClearLOS, int32_t* bSeen, int32_t* bHeard, int32_t bTargetInvisible)
+{
+    Services::PerObjectStorageProxy* pPOS = g_plugin->GetServices()->m_perObjectStorage.get();
+    
+    if (pObserverCreature->m_bPlayerCharacter && pHidingCreature->m_bPlayerCharacter && pHidingCreature->m_nStealthMode)
+    {
+        if (pObserverCreature->GetArea() == pHidingCreature->GetArea())
+        {
+            if (*pPOS->Get<int>(pHidingCreature->m_idSelf, revealKey + "PARTY"))
```

Comment:
I'm not sure this is so well defined. `pPos->Get<int>(...)` will return a `Maybe<int>`, and `*` of that will be the int. But if the value does not exist, dereferencing it with * is undefined.

I think you want:
```C
auto reveal = pPOS->Get<int>(pHidingCreature->m_idSelf, revealKey + "PARTY");
if (reveal && *reveal)
{
    ...
```
And same for other cases